### DISCUSSION TITLE
ERROR building the code with intel compiler

### DISCUSSION THREAD

**Discussion by hamhamm** [modern-fortran/neural-fortran#163] [Q&A]
Hello. Errors came up when i used the intel 2018 compiler (intel-compiler-2018.5.274) to build the code. Could someone help pls?
The building command i used was:

mkdir build
cd build
CC=icc FC=ifort cmake .. -DSERIAL=1
make

And when doing the "make“ process, i got errors like:
.......
Scanning dependencies of target test_reshape_layer
[ 54%] Building Fortran object test/CMakeFiles/test_reshape_layer.dir/test_reshape_layer.f90.o
[ 55%] Linking Fortran executable ../bin/test_reshape_layer
ld: warning: libmpi_mt.so.4, needed by /public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so, not found (try using -rpath or -rpath-link)
../lib/libneural.a(nf_network_submodule.f90.o): In function `nf_network_mp_train_':
nf_network_submodule.f90:(.text+0x97c3): undefined reference to `co_broadcast_'
nf_network_submodule.f90:(.text+0x97d7): undefined reference to `co_broadcast_'
../lib/libneural.a(nf_network_submodule.f90.o): In function `nf_network_mp_update_':
nf_network_submodule.f90:(.text+0x9e44): undefined reference to `co_sum_'
nf_network_submodule.f90:(.text+0x9e52): undefined reference to `co_sum_'
nf_network_submodule.f90:(.text+0x9e6c): undefined reference to `co_sum_'
nf_network_submodule.f90:(.text+0x9e7a): undefined reference to `co_sum_'
../lib/libneural.a(nf_dense_layer_submodule.f90.o): In function `nf_dense_layer_mp_init_':
nf_dense_layer_submodule.f90:(.text+0x3b93): undefined reference to `co_broadcast_'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Barrier'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Send'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Comm_dup'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Put'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Get'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Init'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Free_mem'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Win_unlock'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Type_vector'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Alloc_mem'
/public/software/compiler/intel/intel-compiler-2018.5.274/compiler/lib/intel64/libicaf.so: undefined reference to `MPI_Type_create_struct'
........

**Comment by milancurcic** [comment]
Hi @hamhamm, I don't have access to ifort 2018, but with ifort v2021.10.0, no problems here. Can you check if your ifort supports the `-coarray=single` flag? That flag is activated when you `cmake` with `-DSERIAL=1`, and it should not attempt to link against MPI, which seems to be the cause of the problem here.

**Reply by hamhamm** [reply]
Thank you for your suggestion. It was indeed related to the version of ifort. The old version ifort seems not to work with the -coarray=single flag. After then I managed to use the new ifort 2022, and the code is now compiled successfully.
